src/sagemaker/estimator.py

```diff
@@ -599,10 +601,33 @@ def __init__(
         self.output_kms_key = resolve_value_from_config(
             output_kms_key, TRAINING_JOB_KMS_KEY_ID_PATH, sagemaker_session=self.sagemaker_session
         )
-        self.volume_kms_key = resolve_value_from_config(
-            volume_kms_key,
-            TRAINING_JOB_VOLUME_KMS_KEY_ID_PATH,
-            sagemaker_session=self.sagemaker_session,
+        if instance_type is None or isinstance(instance_type, str):
+            instance_type_for_volume_kms = instance_type
+        elif isinstance(instance_type, ParameterString):
+            instance_type_for_volume_kms = instance_type.default_value
+        else:
+            raise ValueError(f"Bad value for instance type: '{instance_type}'")
+
+        # KMS can only be attached to supported instances
+        use_volume_kms_config = (
+            (instance_type_for_volume_kms and instance_supports_kms(instance_type_for_volume_kms))
+            or instance_groups is not None
+            and any(
+                [
+                    instance_supports_kms(instance_group.instance_type)
+                    for instance_group in instance_groups
+                ]
+            )
+        )
+
+        self.volume_kms_key = (
+            resolve_value_from_config(
+                volume_kms_key,
+                TRAINING_JOB_VOLUME_KMS_KEY_ID_PATH,
+                sagemaker_session=self.sagemaker_session,
+            )
+            if use_volume_kms_config
+            else volume_kms_key
```

Comment:
What are the alternate options to skipping volume kms key default configs that the admin set? If an admin set volume kms keys in the config, is it reasonable for them to expect that it will only be plugged in for only some instance types? Or would they prefer that their users only use instance types that support kms keys?

I think its worth looping in Prateek to discuss whether this is is the right approach

Reply:
I think it's fine to have this conditional logic as we dont want to limit users to use only those instances which support volume kms key.
cc - @prateekmehrotra

Reply:
And I suppose the whole purpose of the Config is to improve user experience. If the customer wants to enforce that KMS keys be used always, the right place for them to strictly enforce that would be through IAM Policies, SCPs, etc. There is no guarantee that populating the Config means those values will be used no matter what.